### ISSUE TITLE
Problem with self in magic methods like `__add__`

### ISSUE THREAD

**Issue by tmke8** [microsoft/pyright#6388]
**Describe the bug**
I noticed the problem with `__add__`, but it also appears when calling a method with `A.f(self)` insead of `self.f()`. These two facts might be related.

**Code or Screenshots**

```python
from typing import Self

class A:
    def __add__(self, other: int) -> Self:
        return self

    def f(self) -> Self:
        return self + 0  # Type "A" cannot be assigned to type "Self@A"

    def g(self, other: int) -> Self:
        return self

    def h(self) -> Self:
        return A.g(self, 0)  # Type "A" cannot be assigned to type "Self@A"
```

**VS Code extension or command-line**
command-line

**Comment by erictraut** [comment]
Thanks for the bug report. 

You've identified two separate issues in your code sample. The first one is definitely a bug, and it will be fixed in the next release.

As for the second issue, I'm not convinced this is a bug. I think pyright is arguably doing the right thing by binding the type `A` to the method `g`. This may seem like premature binding, but it would be important if `A` were a generic class. For example, `A[int].g(self, 0)` would be problematic if pyright didn't bind `A[int]` to `g`. This would result in a false negative. 

So I'm going to leave the current behavior in place for the second issue.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.336, which I just published. It will also be included in a future release of pylance.
